File: interview_questions/paren_permutation.py
res = []
counter = 0

def gen(n, left, right, current):

    global counter
    print(counter)
    counter += 1
    if left == n and right == n:
        res.append(current)
    else:
        if left < n:
            gen(n, left + 1, right, current + "(")
        if left > right:
            gen(n, left, right + 1, current + ")")


gen(3, 0, 0, "")
print(res)

File: interview_questions/paren_permutation.cs
List<string> answer = new List<string>();
void getValidPalindromes(int n, int left, int right, string current)
{
    if (left == n && right == n)
    {
        answer.Add(current);
    }

    if (left <= n)
    {
        getValidPalindromes(n, left + 1, right, current + "(");
    }

    if (right < left)
    {
        getValidPalindromes(n, left, right + 1, current + ")");
    }
}

getValidPalindromes(5, 0, 0, "");
foreach(string a in answer)
{
    Console.WriteLine(a);
}
